src/storage/compression/bitpacking_utils.cpp

```diff
@@ -0,0 +1,203 @@
+
+        if (shiftLeft == 0) {
+            out[0] = static_cast<CompressedType>(in & mask);
+        } else {
+            out[0] |= static_cast<CompressedType>(in << shiftLeft);
+        }
+
+        out[1] = static_cast<CompressedType>((in & mask) >> (compressedFieldSizeBits - shiftLeft));
+        out += 2;
+
+        if (delta + shiftLeft > 2 * compressedFieldSizeBits) {
+            *out = static_cast<CompressedType>(
+                (in & mask) >> (2 * compressedFieldSizeBits - shiftLeft));
+        }
+    }
+
+    else if (delta + shiftLeft >= 3 * compressedFieldSizeBits &&
+             delta + shiftLeft < 4 * compressedFieldSizeBits) {
+        if (shiftLeft == 0) {
+            out[0] = static_cast<CompressedType>(in & mask);
+        } else {
+            out[0] |= static_cast<CompressedType>(in << shiftLeft);
+        }
+
+        out[1] = static_cast<CompressedType>((in & mask) >> (compressedFieldSizeBits - shiftLeft));
+        out[2] =
+            static_cast<CompressedType>((in & mask) >> (2 * compressedFieldSizeBits - shiftLeft));
+        out += 3;
+
+        if (delta + shiftLeft > 3 * compressedFieldSizeBits) {
+            *out = static_cast<CompressedType>(
+                (in & mask) >> (3 * compressedFieldSizeBits - shiftLeft));
+        }
+    }
+
+    else if (delta + shiftLeft >= 4 * compressedFieldSizeBits) {
+        if (shiftLeft == 0) {
+            out[0] = static_cast<CompressedType>(in & mask);
+        } else {
+            out[0] |= static_cast<CompressedType>(in << shiftLeft);
+        }
+        out[1] = static_cast<CompressedType>((in & mask) >> (compressedFieldSizeBits - shiftLeft));
+        out[2] =
+            static_cast<CompressedType>((in & mask) >> (2 * compressedFieldSizeBits - shiftLeft));
+        out[3] =
+            static_cast<CompressedType>((in & mask) >> (3 * compressedFieldSizeBits - shiftLeft));
+        out += 4;
+
+        if (delta + shiftLeft > 4 * compressedFieldSizeBits) {
+            *out = static_cast<CompressedType>(
+                (in & mask) >> (4 * compressedFieldSizeBits - shiftLeft));
+        }
+    }
+}
+
+template<IntegerBitpackingType UncompressedType>
+    requires(sizeof(UncompressedType) >= sizeof(uint32_t))
+void SingleValuePacker<UncompressedType>::unpackSingle(const uint32_t* __restrict& srcCursor,
+    UncompressedType* __restrict dst, uint16_t bitWidth, size_t srcOffset) {
+    const size_t shiftRight = srcOffset * bitWidth % (sizeof(uint32_t) * 8);
+    unpackSingleImpl(srcCursor, dst, bitWidth, shiftRight);
+}
+
+template<IntegerBitpackingType UncompressedType>
+    requires(sizeof(UncompressedType) < sizeof(uint32_t))
+void SingleValuePacker<UncompressedType>::unpackSingle(const uint8_t* __restrict& srcCursor,
+    UncompressedType* __restrict dst, uint16_t bitWidth, size_t srcOffset) {
+    const size_t shiftRight = srcOffset * bitWidth % (sizeof(uint8_t) * 8);
+    unpackSingleImpl(srcCursor, dst, bitWidth, shiftRight);
+}
```

Comment:
These overloads could probably be removed (since we need to use functions like `castCompressedCursor()` anyways to determine if the compressed field size is 32 or 8 bits) but removing this layer would lose the strictness of mapping each uncompressed type to a compressed type.

Reply:
removed